bus-mapping/src/evm/opcodes/calldataload.rs

```diff
@@ -0,0 +1,207 @@
+use crate::{
+    circuit_input_builder::{CircuitInputStateRef, ExecStep},
+    operation::{CallContextField, CallContextOp, RW},
+    Error,
+};
+use eth_types::{GethExecStep, U256};
+
+use super::Opcode;
+
+#[derive(Clone, Copy, Debug)]
+pub(crate) struct Calldataload;
+
+impl Opcode for Calldataload {
+    fn gen_associated_ops(
+        state: &mut CircuitInputStateRef,
+        geth_steps: &[GethExecStep],
+    ) -> Result<Vec<ExecStep>, Error> {
+        let geth_step = &geth_steps[0];
+        let mut exec_step = state.new_step(geth_step)?;
+
+        // fetch the top of the stack, i.e. offset in calldata to start reading 32-bytes
+        // from.
+        let offset = geth_step.stack.nth_last(0)?;
+
+        state.push_stack_op(
+            &mut exec_step,
+            RW::READ,
+            geth_step.stack.nth_last_filled(0),
+            offset,
+        )?;
+
+        state.push_op(
+            &mut exec_step,
+            RW::READ,
+            CallContextOp {
+                call_id: state.call()?.call_id,
+                field: CallContextField::TxId,
+                value: state.tx_ctx.id().into(),
+            },
+        );
+
+        if !state.call()?.is_root {
```

Comment:
You're missing an `else` clause here for the `root` case where we Read `CallDataLength` only.
See: https://github.com/appliedzkp/zkevm-specs/blob/master/src/zkevm_specs/evm/execution/calldataload.py#L17-L20

Reply:
Since the `call_data_length` in the `root` case comes from Transaction context, we don't need to populate the RW table. Only in the `internal` case it is read from the RW table's CallContext.

Reply:
Ohhh that's right!!!! Good one!!!